modules/appeals_api/app/services/appeals_api/uploads/document_validation.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'pdf_info'
+
+module AppealsApi
+  module Uploads
+    class DocumentValidation
+      def initialize(upload)
+        @document = upload[:document]
+      end
+
+      def validate
+        # TODO: log anything here if errors raised or log in the controller if fails validation?
+        raise UploadValidationError, non_pdf_error unless document_is_pdf?
+        raise UploadValidationError, max_dimension_error unless valid_page_dimensions?
+        raise UploadValidationError, file_size_error unless valid_file_size?
```

Comment:
Let's try to make sure that each of these checks only reads the contents into memory as few times as possible - either by caching the contents in an instance variable or only reading the minimum of the file necessary to get the job done for each validation. Most of these checks are likely doing this already, but we should verify that.

If some of these require the full load of the file into memory, let's rearrange these validations in order of "expense", where the leanest come first and the most memory intensive are last.

Reply:
OK I'll dig into this.